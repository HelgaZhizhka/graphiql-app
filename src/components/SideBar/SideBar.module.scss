.root {
  position: relative;
}

.button {
  position: absolute;
  top: 100px;
  left: -60px;
  z-index: var(--z-index-sidebar);
  transform: rotate(-90deg);
}

.panel {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  width: 300px;
  background-color: var(--component-bg);
  transform: translateX(100%);
  transition: transform 0.3s ease-in-out;
  // overflow-y: auto;
  z-index: var(--z-index-sidebar);
}

.open {
  transform: translateX(0);
}
