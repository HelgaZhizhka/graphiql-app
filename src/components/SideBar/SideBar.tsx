import { useState } from 'react';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import CloseIcon from '@mui/icons-material/Close';
import IconButton from '@mui/material/IconButton';

import { SchemaExplorer } from '@/components/SchemaExplorer';
import { CenteredTypography } from '../CenteredTypography';
import styles from './SideBar.module.scss';

const SideBar: React.FC = () => {
  const [open, setOpen] = useState(false);

  const handleClick = () => {
    setOpen(!open);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <div className={open ? `${styles.root} ${styles.open}` : styles.root}>
      <Button className={styles.button} variant="contained" color="primary" onClick={handleClick}>
        Schema
      </Button>
      <div className={styles.panel}>
        {open && (
          <>
            <div className={styles.panelHeader}>
              <CenteredTypography>Documentation Explorer</CenteredTypography>
              <IconButton
                aria-label="close"
                color="primary"
                className={styles.close}
                onClick={handleClose}
              >
                <CloseIcon />
              </IconButton>
            </div>
            <div className={styles.panelBody}>
              <Typography mb={2}>
                A GraphQL schema provides a root type for each kind of operation.
              </Typography>
              <SchemaExplorer />
            </div>
          </>
        )}
      </div>
    </div>
  );
};

export default SideBar;
