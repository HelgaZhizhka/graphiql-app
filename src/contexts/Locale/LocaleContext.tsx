import { ReactNode, useReducer } from 'react';
import {
  LOCALE_STRINGS,
  LocalState,
  LocaleAction,
  LocaleContext,
  Props,
  initialState,
} from './constants';

const reducer = (state: LocalState, action: LocaleAction) => {
  switch (action.type) {
    case 'CHANGE_LOCALE': {
      localStorage.setItem('lang', action.payload.region);
      return {
        ...state,
        strings: LOCALE_STRINGS[action.payload.region],
      };
    }
    default:
      return state;
  }
};

export const LocaleProvider: React.FC<{ children: ReactNode }> = ({ children }: Props) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const contextValue = {
    state,
    dispatch,
  };

  return <LocaleContext.Provider value={contextValue}>{children}</LocaleContext.Provider>;
};
