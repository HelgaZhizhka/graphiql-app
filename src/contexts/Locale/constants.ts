import enStrings from '@/locales/en.json';
import ruStrings from '@/locales/ru.json';
import errorsEn from '@/locales/errorsEn.json';
import errorsRu from '@/locales/errorsRu.json';

export const REGIONS = {
  RU: 'RU',
  EN: 'EN',
};

export const LOCALE_STRINGS = {
  [REGIONS.EN]: { ...enStrings, ...errorsEn },
  [REGIONS.RU]: { ...ruStrings, ...errorsRu },
};

export const getLang = () => localStorage.getItem('lang') ?? REGIONS.EN;
export const lang = getLang();

export const initialState = {
  strings: LOCALE_STRINGS[lang],
};

export type vocab = typeof initialState.strings;
