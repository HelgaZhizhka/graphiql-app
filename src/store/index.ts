import { configureStore } from '@reduxjs/toolkit';

import errorReducer from './slices/errorSlice';

const store = configureStore({
  reducer: {
    error: errorReducer,
  },
});

export default store;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
