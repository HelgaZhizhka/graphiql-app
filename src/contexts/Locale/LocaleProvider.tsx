import { useContext } from 'react';

import { LocaleContext } from './LocaleContext';

export const useLocale = () => {
  const contextValue = useContext(LocaleContext);

  if (!contextValue) {
    throw new Error('useLocale must be used within a LocaleProvider');
  }

  return contextValue;
};
