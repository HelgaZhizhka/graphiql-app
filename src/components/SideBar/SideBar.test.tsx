import '@testing-library/jest-dom';
import { screen, render, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '@/store';
import { LocaleProvider } from '@/contexts/Locale/LocaleContext';
import SideBar from './SideBar';

describe('Testing SideBar Component', () => {
  it('renders component with closed panel by default', () => {
    const { container } = render(
      <Provider store={store}>
        <LocaleProvider>
          <SideBar />
        </LocaleProvider>
      </Provider>
    );

    const panel = container.getElementsByClassName('panel');
    expect(panel).toHaveLength(1);
  });

  it('opens panel when "Schema" button is clicked', async () => {
    render(
      <Provider store={store}>
        <LocaleProvider>
          <SideBar />
        </LocaleProvider>
      </Provider>
    );

    const button = screen.getByText(/Open/i);
    fireEvent.click(button);

    await waitFor(() => {
      const panel = screen.getByText(/Documentation Explorer/i);
      expect(panel).toBeInTheDocument();
    });
  });

  it('closes panel when "Close" button is clicked', () => {
    render(
      <Provider store={store}>
        <LocaleProvider>
          <SideBar />
        </LocaleProvider>
      </Provider>
    );

    const button = screen.getByText(/Open/i);
    fireEvent.click(button);
    const closeButton = screen.getByLabelText('close');
    fireEvent.click(closeButton);
    const panel = screen.queryByText(/Documentation Explorer/i);
    expect(panel).not.toBeInTheDocument();
  });

  it('displays correct text in the panel', async () => {
    render(
      <Provider store={store}>
        <LocaleProvider>
          <SideBar />
        </LocaleProvider>
      </Provider>
    );

    const button = screen.getByText(/Open/i);
    fireEvent.click(button);

    await waitFor(() => {
      const textInPanel = screen.getByText(
        'A GraphQL schema provides a root type for each kind of operation.'
      );
      expect(textInPanel).toBeInTheDocument();
    });
  });
});
