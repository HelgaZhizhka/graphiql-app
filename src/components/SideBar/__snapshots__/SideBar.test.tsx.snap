// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SideBar Component renders component with closed panel by default 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium button css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Schema
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="panel"
    />
  </div>
</DocumentFragment>
`;
